src/planner/partialize.c

```diff
@@ -185,6 +197,652 @@ partialize_agg_paths(RelOptInfo *rel)
+	{
+		Path *append_path = lfirst(lc);
+		double total_groups = append_path->rows * append_path->parallel_workers;
+
+		Path *gather_path = (Path *) create_gather_path(root,
+														partially_grouped_rel,
+														append_path,
+														partially_grouped_rel->reltarget,
+														NULL,
+														&total_groups);
+		add_path(partially_grouped_rel, (Path *) gather_path);
+	}
+}
+
+/*
+ * Get the best total path for aggregation. Prefer chunk append paths if we have one, otherwise
+ * return the cheapest_total_path;
+ */
+static Path *
+get_best_total_path(RelOptInfo *output_rel)
+{
+	ListCell *lc;
+	foreach (lc, output_rel->pathlist)
+	{
+		Path *path = lfirst(lc);
+
+		if (ts_is_chunk_append_path(path))
+			return path;
+	}
+
+	return output_rel->cheapest_total_path;
+}
+
+/*
+ * Check if this path belongs to a plain or sorted aggregation
+ */
+static bool
+is_plain_or_sorted_agg_path(Path *path)
+{
+	List *subpaths = get_subpaths_from_append_path(path, true);
+
+	Ensure(subpaths != NIL, "Unable to determine aggregation type");
+
+	Path *subpath = linitial(subpaths);
+
+	if (IsA(subpath, AggPath))
+	{
+		AggPath *agg_path = castNode(AggPath, linitial(subpaths));
+		Assert(agg_path->aggstrategy == AGG_SORTED || agg_path->aggstrategy == AGG_PLAIN ||
+			   agg_path->aggstrategy == AGG_HASHED);
+		return agg_path->aggstrategy == AGG_SORTED || agg_path->aggstrategy == AGG_PLAIN;
+	}
+
+	return is_plain_or_sorted_agg_path(subpath);
+}
+
+/*
+ * Replan the aggregation and create a partial aggregation at chunk level and finalize the
+ * aggregation on top of an append node.
+ *
+ * The functionality is inspired by PostgreSQL's create_partitionwise_grouping_paths() function
+ *
+ * Generated aggregation paths:
+ *
+ * Finalize Aggregate
+ *   -> Append
+ *      -> Partial Aggregation
+ *        - Chunk 1
+ *      ...
+ *      -> Append of partially compressed chunk 2
+ *         -> Partial Aggregation
+ *             -> Scan on uncompressed part of chunk 2
+ *         -> Partial Aggregation
+ *             -> Scan on compressed part of chunk 2
+ *      ...
+ *      -> Partial Aggregation N
+ *        - Chunk N
+ */
+void
+ts_pushdown_partial_agg(PlannerInfo *root, Hypertable *ht, RelOptInfo *input_rel,
+						RelOptInfo *output_rel, void *extra)
+{
+	Query *parse = root->parse;
+
+	/* We are only interested in hypertables */
+	if (ht == NULL || hypertable_is_distributed(ht))
+		return;
+
+	/* Perform partial aggregation planning only if there is an aggregation is requested */
+	if (!parse->hasAggs)
+		return;
+
+	/* Groupting sets are not supported by the partial aggregation pushdown */
+	if (parse->groupingSets)
+		return;
+
+	/* Don't replan aggregation if we already have a MinMaxAggPath (e.g., created by
+	 * ts_preprocess_first_last_aggregates) */
+	if (has_min_max_agg_path(output_rel))
+		return;
+
+	/* Is sorting possible ? */
+	bool can_sort = grouping_is_sortable(parse->groupClause) &&
```

Comment:
How will these GUCs interact with the PG setting `enable_partitionwise_aggregate`, if at all? Will you be able to disable _our_ partitionwise aggregation with the PG setting or are these completely standalone?

Just wondering if this could cause confusion on how to enable/disable the functionality.

Reply:
I replaced the two GUCs by the new GUC `timescaledb.enable_partitionwise_aggregation`. If this GUC is disabled, our pushdown code is never invoked and we keep the planned PostgreSQL aggregation paths (and the aggregation paths created by other TimescaleDB optimizations).